src/app-layer-dnp3.c

```diff
@@ -712,7 +712,7 @@ static int DNP3DecodeApplicationObjects(DNP3Transaction *tx, const uint8_t *buf,
                 }
                 object->start = buf[offset++];
                 object->stop = buf[offset++];
-                object->count = object->stop - object->start + 1;
+                object->count = object->stop >= object->start ? object->stop - object->start + 1 : 0;
```

Comment:
I can not provide such pcap since it is from the customer. However, i can explain my point and a way how to prepare a reproducer.
First of all, we can not rely on healthy data content. It may be created e.g. by malicious or malfunctioned software/device. Offsets, data size, etc. taken from l7 header must be always validated.

The most dangerous part of code is fixed here:

@@ -1022,6 +1031,10 @@ static void DNP3HandleUserDataResponse(DNP3State *dnp3, const uint8_t *input,
     tx->response_done = 1;
 
     offset = sizeof(DNP3ApplicationHeader) + sizeof(DNP3InternalInd);
+
+    if (tx->response_buffer_len < offset)
+        return;
+
     if (DNP3DecodeApplicationObjects(tx, tx->response_buffer + offset,
             tx->response_buffer_len - offset,
             &tx->response_objects)) {

And the similar in the DNP3HandleUserDataRequest
@@ -919,6 +922,9 @@ static void DNP3HandleUserDataRequest(DNP3State *dnp3, const uint8_t *input,

Nobody guarantees a valid response with some objects contained. The protocol specification does not allow this, however it may be a result of attack/malfunction.

As a result, DNP3DecodeApplicationObjects gets a pointer to some random data, so this check in the DNP3HandleUserDataResponse is absolutely required. Even it passed, we have to validate the count of objects calculated inside DNP3DecodeApplicationObjects. Nobody guarantees a valid content of the data.

The easiest way to reproduce the issue with some valid communication is to use wrong direction detection if the analysis is started from midstream. So please enable stream.midstream in the suricata.yaml and prepare a pcap that starts from a response (src port 20000) but also contains a CONFIRM packet from the client. The confirm packet is treated as a response; a pointer to some random data is passed to the DNP3DecodeApplicationObjects and some random number of objects is finally calculated.

The example also indicates that currently, dnp3 parser does not detect the proper direction of the stream. It's a different issue.

Well, some other checks I introduced are paranoid. I added sanity checks at all places where pointer arithmetic looks unsafe.

Reply:
Thanks for the explanation.

How come the call to `DNP3HasUserData` does not work and ensure enough data in `DNP3HandleUserDataResponse` ?

Reply:
> I can not provide such pcap since it is from the customer. However, i can explain my point and a way how to prepare a reproducer.

Could you create a unit-test that triggers these conditions?

Reply:
> How come the call to `DNP3HasUserData` does not work and ensure enough data in `DNP3HandleUserDataResponse` ?

DNP3HasUserData only verifies the internal consistency of the header. It does not take into account actually available amount of data at all.


> Could you create a unit-test that triggers these conditions?
I'll try. I hope it's possible.

Reply:
> The example also indicates that currently, dnp3 parser does not detect the proper direction of the stream. It's a different issue.

Indeed, I created https://redmine.openinfosecfoundation.org/issues/3772 to track this

Reply:
My understanding is that
- condition `if (input_len < sizeof(DNP3LinkHeader)) ` at line 1060 ensures there is enough data in the buffer which can contain many frames
- condition `if (input_len < frame_len)` 20 lines below ensure the frame is complete in the current buffer
- condition `if (!DNP3IsUserData(header))` ensures the frame has user data

So, all these conditions together, and the fact that `DNP3CalculateLinkLength` computes the right value, seem to make it ok to use `offset = sizeof(DNP3ApplicationHeader) + sizeof(DNP3InternalInd);` in the current frame

Am I mistaken somewhere ?

This seems indeed a bit fragile, and the use of more `DEBUG_VALIDATE_BUG_ON` is indeed welcome to ensure the buffers are not overrrun